common/scala/src/main/scala/org/apache/openwhisk/core/database/mongodb/MongoDBArtifactStore.scala

```diff
@@ -0,0 +1,661 @@
+ * @param documentHandler helper class help to simulate the designDoc of CouchDB
+ * @param viewMapper helper class help to simulate the designDoc of CouchDB
+ */
+class MongoDBArtifactStore[DocumentAbstraction <: DocumentSerializer](client: MongoClient,
+                                                                      dbName: String,
+                                                                      collName: String,
+                                                                      documentHandler: DocumentHandler,
+                                                                      viewMapper: MongoDBViewMapper,
+                                                                      val inliningConfig: InliningConfig,
+                                                                      val attachmentStore: Option[AttachmentStore])(
+  implicit system: ActorSystem,
+  val logging: Logging,
+  jsonFormat: RootJsonFormat[DocumentAbstraction],
+  val materializer: ActorMaterializer,
+  docReader: DocumentReader)
+    extends ArtifactStore[DocumentAbstraction]
+    with DocumentProvider
+    with DefaultJsonProtocol
+    with AttachmentSupport[DocumentAbstraction] {
+
+  import MongoDBArtifactStore._
+
+  protected[core] implicit val executionContext = system.dispatcher
+
+  private val mongodbScheme = "mongodb"
+  val attachmentScheme: String = attachmentStore.map(_.scheme).getOrElse(mongodbScheme)
+
+  private val database = client.getDatabase(dbName)
+  private val collection = getCollectionAndCreateIndexes
+  private val gridFSBucket = GridFSBucket(database, collName)
+
+  private val jsonWriteSettings = JsonWriterSettings.builder().outputMode(JsonMode.RELAXED).build
+
+  // MongoDB doesn't support using `$` as the first char of field name, so below two fields needs to be encoded first
+  private val fieldsNeedEncode = Seq("annotations", "parameters")
+
+  override protected[database] def put(d: DocumentAbstraction)(implicit transid: TransactionId): Future[DocInfo] = {
+    val asJson = d.toDocumentRecord
+
+    val id: String = asJson.fields.getOrElse("_id", JsString.empty).convertTo[String].trim
+    require(!id.isEmpty, "document id must be defined")
+
+    val (old_rev, rev) = revisionCalculate(asJson)
+    val docinfoStr = s"id: $id, rev: $rev"
+    val start =
+      transid.started(this, LoggingMarkers.DATABASE_SAVE, s"[PUT] '$collName' saving document: '$docinfoStr'")
+
+    val encodedData = encodeFields(fieldsNeedEncode, asJson)
+
+    val data = JsObject(
+      encodedData.fields + (_computed -> documentHandler.computedFields(asJson)) + ("_rev" -> rev.toJson))
+
+    val filters =
+      if (rev.startsWith("1-")) {
+        // for new document, we should get no matched document and insert new one
+        // if there is a matched document, that one with no _rev filed will be replaced
+        // if there is a document with the same id but has an _rev field, will return en E11000(conflict) error
```

Comment:
Is it possible for MongoDB to include `_rev` field?

Reply:
sorry, I didn't get it?

Reply:
ok I got confused.
According to the description in this PR, MongoDB doesn't have this field.

> The data scheme in MongoDB is just like CouchDB except that MongoDB doesn't have a _rev field, below is the data of whisk.system/invokerHealthTestAction0 in CouchDB and MongoDB:

So I thought we can't use a field name `_rev` for some reason, but it seems not.
While CouchDB manages `_rev` value, this will manually manage revision using revisionCalculator.
https://github.com/apache/openwhisk/pull/4963/files/68120f2170dc9f9b53361ab0cb51c4e9458dbe29#diff-f5da2a3c01f4ecedf28b8c10af6bb42c71ce4ad7f66424497f81523c17e33469R572

So the field `_rev` is also stored in MongoDB along with other fields.

Reply:
yes, CouchDb will generate `_rev` automatically, while MongoDB should set this field manually